test/unit/blueprintUtils-test.js

```diff
@@ -139,24 +142,82 @@ describe('blueprintUtils', () => {
 
         yup!\
 `;
-        drafter.parse(blueprint, options, (err, parseResult) => {
-          if (err) { return done(err); }
-
-          const annotations = (Array.from(parseResult.content).filter(node => node.element === 'annotation'));
-          warnings = annotations.length;
+      const annotations = [
+        {
+          element: 'annotation',
+          meta: { classes: ['warning'] },
+          attributes: {
+            code: 10,
+            sourceMap: [
+              {
+                element: 'sourceMap',
+                content: [[70, 23], [95, 24]]
+              }
+            ]
+          },
+          content:
+            'message-body asset is expected to be a pre-formatted code ' +
+            'block, every of its line indented by exactly 8 spaces or 2 tabs'
+        },
+        {
+          element: 'annotation',
+          meta: { classes: ['warning'] },
+          attributes: {
+            code: 10,
+            sourceMap: [
+              {
+                element: 'sourceMap',
+                content: [[168, 39]]
+              }
+            ]
+          },
+          content:
+            'headers is expected to be a pre-formatted code block, every ' +
+            'of its line indented by exactly 12 spaces or 3 tabs'
+        },
+        {
+          element: 'annotation',
+          meta: { classes: ['warning'] },
+          attributes: {
+            code: 10,
+            sourceMap: [
+              {
+                element: 'sourceMap',
+                content: [[224, 33]]
+              }
+            ]
+          },
+          content:
+            'message-body asset is expected to be a pre-formatted code ' +
+            'block, every of its line indented by exactly 12 spaces or 3 tabs'
+        },
+        {
+          element: 'annotation',
+          meta: { classes: ['warning'] },
+          attributes: {
+            code: 10,
+            sourceMap: [
+              {
+                element: 'sourceMap',
+                content: [[302, 12], [318, 12], [334, 14]]
+              }
+            ]
+          },
+          content:
+            'message-body asset is expected to be a pre-formatted code ' +
+            'block, every of its line indented by exactly 8 spaces or 2 tabs'
+        }
+      ];
 
-          for (const annotation of annotations) {
-            let location = [];
-            for (const sourceMap of annotation.attributes.sourceMap) {
-              location = location.concat(sourceMap.content);
-            }
-            allRanges.push(blueprintUtils.warningLocationToRanges(location, blueprint));
-          }
-          done();
-        });
-      });
+      for (const annotation of annotations) {
+        let location = [];
+        for (const sourceMap of annotation.attributes.sourceMap) {
+          location = location.concat(sourceMap.content);
```

Comment:
What about using [`flatMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap#Examples) instead of imperitive loop/concat?

```js
const location = annotation.attributes.sourceMap.flatMap((sourceMap) => sourceMap.content);
```

Reply:
Oh this is just unindented. I don't like the code, I could make a comment about every single line of it myself, but I don't want to go into rewriting it now, especially given I think I should be able to drop the whole file soon.